lib/python/picongpu/utils/picongpu_memory_calculator.py

```diff
@@ -0,0 +1,363 @@
+                [16, 16, 1])  # \TODO make this more generic
+            local_cells = (n_x + supercell_size[0] * 2 * guard_size_supercells[
+                0]) * (n_y + supercell_size[1] * 2 * guard_size_supercells[1])
+
+            # cells around core-border region due to particle shape
+            double_buffer_cells = (n_x + pso) * (n_y + pso) - n_x * n_y
+        elif sim_dim == 3:
+            # super cell size in cells in x, y, z
+            # \TODO make this more generic
+            supercell_size = np.array([8, 8, 4])
+            local_cells = (
+                n_x + supercell_size[0] * 2 *
+                guard_size_supercells[0]) \
+                * (n_y + supercell_size[1] * 2 *
+                   guard_size_supercells[1]) \
+                * (n_z + supercell_size[2] * 2 *
+                   guard_size_supercells[2])
+
+            # cells around core-border region due to particle shape
+            double_buffer_cells = (n_x + pso) * (n_y + pso) * (n_z + pso) \
+                - n_x * n_y * n_z
+        else:
+            raise ValueError(
+                "PIConGPU only runs in either 2D or 3D: ",
+                sim_dim,
+                " =/= {2, 3}")
+
+        # size of a data entry in bytes
+        data_size = np.float32().itemsize
+        # number of fields: 3 * 3 = x,y,z for E,B,J
+        num_fields = 3 * 3 + field_tmp_slots
+        # double buffer memory
+        double_buffer_mem = double_buffer_cells * num_fields * data_size
+
+        req_mem = data_size * num_fields * local_cells + double_buffer_mem
+        return req_mem
+
+    def mem_req_by_particles(
+            self,
+            target_n_x=None,
+            target_n_y=None,
+            target_n_z=None,
+            num_additional_attributes=0,
+            particles_per_cell=2
+    ):
+        """
+        Memory reserved for all particles of a species on a GPU.
+        We currently neglect the constant species memory.
+
+        Parameters
+        ----------
+
+        target_n_x : int
+            number of cells in x direction containing the target
+        target_n_y : int
+            number of cells in y direction containing the target
+        target_n_z : int
+            number of cells in z direction containing the target
+        num_additional_attributes : int
+            number of additional attributes like e.g. ``boundElectrons``
+        particles_per_cell : int
+            number of particles of the species per cell
+
+        Returns
+        -------
+
+        req_mem : int
+            required memory {unit: bytes} per GPU and species
+        """
+
+        if target_n_x is None:
+            target_n_x = self.n_x
+        if target_n_y is None:
+            target_n_y = self.n_y
+        if target_n_z is None:
+            target_n_z = self.n_z
+
+        # memory required by the standard particle attributes
+        standard_attribute_mem = np.array([
+            3 * 4,  # momentum
+            3 * 4,  # position
```

Comment:
is this not depending on whether the simulation is in 2D or 3D?

Reply:
for position: yes - for momentum: no

will fix it